[rai] Fix extraction for recent raiplay.it updates

iamleot opened the pull request:
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

- Introduce _BASE_URL in RaiBaseIE class so it could be reused as base for the
  several subextractors.
- Remove first test of RaiPlayIE: it is no longer available
- Make RaiPlayIE extension-agnostic (passing possible `.json' URLs is now
  supported too)
- Adjust RaiPlayLiveIE to recent raiplay.it updates.  Passing it as
  `url_transparent' is no longer supported (there is no longer an accessible
  ContentItem)
- Adjust RaiPlayPlaylistIE to recent raiplay.it updates and instruct it about
  ContentSet-s.
- Update a RaiIE test and remove two tests that are no longer availables

This fix issue #22923, #22906 and supersedes #23006 and #23040.

iamleot commented:
JFTR, it seems that the all failed checks are due the Travis CI build that was cancelled (so the code touched by the PR was actually not tested).

remitamine commented on youtube_dl/extractor/rai.py:
```diff
@@ -240,20 +222,32 @@ class RaiPlayLiveIE(RaiBaseIE):
+            'description': media.get('description'),
+            'uploader': strip_or_none(media.get('channel')),
+            'creator': strip_or_none(media.get('editor')),
+            'duration': parse_duration(video.get('duration')),
```
is there a URL where the `duration` is available.

iamleot replied:
There should be... I will try to find one and add it as a test, thanks!

iamleot replied:
Yes, the one in the test (<http://www.raiplay.it/video/2014/04/Report-del-07042014-cb27157f-9dd0-4aee-b788-b1f67643a391.html>) has it.

remitamine replied:
this URL you have provided is not related to this part of the code(`RaiPlayLiveIE`).

iamleot replied:
Whoops, I accidentally confused them with RaiPlayIE, thanks!  I have just double-checked all such URL via <https://www.raiplay.it/dirette/> and none of them have an `alt_title` or `duration` so I removed both of them. This should be addressed by https://github.com/ytdl-org/youtube-dl/pull/27077/commits/2ea4cfb5f8d7a882c2e5c7aedd60af18a36881e5

remitamine commented on youtube_dl/extractor/rai.py:
```diff
@@ -262,29 +256,33 @@ class RaiPlayPlaylistIE(InfoExtractor):
 
-        title = self._html_search_meta(
-            ('programma', 'nomeProgramma'), webpage, 'title')
-        description = unescapeHTML(self._html_search_meta(
-            ('description', 'og:description'), webpage, 'description'))
+        content_sets = [s['id'] for b in media['blocks'] for s in b['sets']]
```
https://github.com/ytdl-org/youtube-dl#mandatory-and-optional-metafields

iamleot replied:
Thanks!  I will readd `title`!

iamleot replied:
I have actually double-checked that and all mandatory metafields seems available... Can you please elaborate what's the problem with the quoted diff?

remitamine replied:
> Any field apart from the aforementioned ones are considered optional. That means that extraction should be tolerant to situations when sources for these fields can potentially be unavailable (even if they are always available at the moment) and future-proof in order not to break the extraction of general purpose mandatory fields.

- the playlist title and description are not mandatory.
- extraction should not fail if any of the fields `blocks`, `sets`, and `id` are not available.

iamleot replied:
Thanks for clarifying, I will adjust that!

iamleot replied:
This should be addressed by https://github.com/ytdl-org/youtube-dl/pull/27077/commits/1ac644a2a06e84c116e2c21d21a842251b8aeb25

remitamine commented on youtube_dl/extractor/rai.py:
```diff
@@ -262,29 +256,33 @@ class RaiPlayPlaylistIE(InfoExtractor):
-                video_url, ie=RaiPlayIE.ie_key(),
-                video_id=RaiPlayIE._match_id(video_url)))
+        for cs in content_sets:
+            medias = self._download_json(
+                '%s/%s.json' % (urljoin(RaiBaseIE._BASE_URL, 'programmi/' + playlist_id), cs),
+                cs, 'Downloading content set JSON')
+            for m in medias['items']:
+                video_url = urljoin(url, m['path_id'])
```
should not fail if the extraction of one set or item is not possible.

iamleot replied:
This should be addressed by https://github.com/ytdl-org/youtube-dl/pull/27077/commits/12c3dae236c0ac5fdeb9ef97e07ac75699c61242

remitamine commented on youtube_dl/extractor/rai.py:
```diff
@@ -240,20 +222,32 @@ class RaiPlayLiveIE(RaiBaseIE):
 
-        webpage = self._download_webpage(url, display_id)
+        media = self._download_json(
+            '%s.json' % urljoin(RaiBaseIE._BASE_URL, 'dirette/' + display_id),
```
capture the base URL in `_VALID_URL` regular expression and then add `.json` suffix(same for `RaiPlayPlaylistIE`).

iamleot replied:
This should be addressed by https://github.com/ytdl-org/youtube-dl/pull/27077/commits/191d41a86a3ff846f132c7e2b97f4db6525bacb4

iamleot commented on youtube_dl/extractor/rai.py:
```diff
@@ -166,10 +148,11 @@ class RaiPlayIE(RaiBaseIE):
     def _real_extract(self, url):
-        url, video_id = re.match(self._VALID_URL, url).groups()
+        mobj = re.match(self._VALID_URL, url)
+        url, base, video_id, ext = mobj.group('url', 'base', 'id', 'ext')
 
         media = self._download_json(
-            url.replace('.html', '.json'), video_id, 'Downloading video JSON')
+            '%s%s.json' % (base, video_id), video_id, 'Downloading video JSON')
```
It wouldn't catch the `.json` extension URLs.

iamleot commented on youtube_dl/extractor/rai.py:
```diff
@@ -122,27 +123,8 @@ def _extract_subtitles(url, subtitle_url):
 
 class RaiPlayIE(RaiBaseIE):
-    _VALID_URL = r'(?P<url>https?://(?:www\.)?raiplay\.it/.+?-(?P<id>%s)\.html)' % RaiBaseIE._UUID_RE
+    _VALID_URL = r'(?P<url>(?P<base>https?://(?:www\.)?raiplay\.it/.+?-)(?P<id>%s)(?P<ext>\.(?:html|json)))' % RaiBaseIE._UUID_RE
```
I will remove unused ones, thanks!

iamleot replied:
This should be addressed by https://github.com/ytdl-org/youtube-dl/pull/27077/commits/db01844f327b081e1921756aeae91deec0d5d747

remitamine commented on youtube_dl/extractor/rai.py:
```diff
@@ -238,53 +219,79 @@ class RaiPlayLiveIE(RaiBaseIE):
+                    content_sets.append(s['id'])
 
         entries = []
-        for mobj in re.finditer(
-                r'<a\b[^>]+\bhref=(["\'])(?P<path>/raiplay/video/.+?)\1',
-                webpage):
-            video_url = urljoin(url, mobj.group('path'))
-            entries.append(self.url_result(
-                video_url, ie=RaiPlayIE.ie_key(),
-                video_id=RaiPlayIE._match_id(video_url)))
+        for cs in content_sets:
```
there is no need to collect the `content_sets` and then iterate over them, you can process them directly.

iamleot replied:
Thanks!  I have refactored it and this should be fixed via https://github.com/ytdl-org/youtube-dl/pull/27077/commits/db472dd61bf49268727470be6ec40ee129d2bca1

remitamine commented on youtube_dl/extractor/rai.py:
```diff
@@ -238,53 +219,79 @@ class RaiPlayLiveIE(RaiBaseIE):
 
 class RaiPlayPlaylistIE(InfoExtractor):
-    _VALID_URL = r'https?://(?:www\.)?raiplay\.it/programmi/(?P<id>[^/?#&]+)'
+    _VALID_URL = r'(?P<base>https?://(?:www\.)?raiplay\.it/)programmi/(?P<id>[^/?#&]+)'
```
the part you're matching is static and won't change, I meant to match to the end of the URL(the same for `RaiPlayLiveIE`).

iamleot replied:
Sorry but then I don't get it.  Please elaborate further.

iamleot replied:
Do you meant to add possible trailing `/` or possible trailing extension (`.html` and/or `.json` if any)?

remitamine replied:
```suggestion
    _VALID_URL = r'(?P<base>https?://(?:www\.)?raiplay\.it/programmi/(?P<id>[^/?#&]+))'
```

iamleot replied:
Thanks!

This should be addressed by https://github.com/ytdl-org/youtube-dl/pull/27077/commits/65600a1efd2891995a5fb6d96d32c77969b1a3ef

iamleot commented:
If/when that looks good to be merged, please let me know that I will rebase and squash all commits in a single one and adjust commit message accordingly (there were something that I added initially but then amended based on review by @remitamine!).

Thanks!

remitamine commented:
you can squash the commits and I will base the rest of the changes on this.

iamleot commented:
Thanks!  I have just squashed all the commits to the only relevant one and updated the commit message.

If there is anything else to do or if possible testing (also in the future) is needed do not hesitate to contact me!

iamleot commented:
remitamine writes:
> Merged #27077 into master.

Thanks!
